oap-storage-mongo/src/main/java/oap/storage/mongo/MongoPersistence.java

```diff
@@ -24,5 +24,177 @@
+        log.debug( "loading data from {}", namespace );
+        Threads.synchronously( lock, () -> {
+            final Consumer<Metadata<T>> cons = metadata -> {
+                var id = storage.identifier.get( metadata.object );
+                storage.data.put( id, metadata );
+            };
+            log.info( "Load {} documents from [{}] Mongo namespace", collection.countDocuments(), namespace );
+            collection.find().forEach( cons );
+        } );
+        log.info( storage.data.size() + " object(s) loaded." );
+    }
+
+    @Override
+    public void updated( T object, boolean added ) {
+        Threads.synchronously( lock, () -> {
+            var id = storage.identifier.get( object );
+            Metadata<T> metadata = storage.data.get( id );
+            if (added) {
+                collection.insertOne( metadata );
+            } else {
+                collection.replaceOne( eq( "_id", id ), metadata, REPLACE_OPTIONS_UPSERT );
+            }
+        } );
+    }
+
+    @Override
+    public void updated( Collection<T> objects ) {
+        objects.forEach( object -> updated( object, false ) );
+
+//        TODO: replace with collection.updateMany()
+//         Create BsonDocument of metadata values
+//        Map<String, Metadata<T>> data = objects.stream()
+//            .map( object -> storage.identifier.get( object ) )
+//            .collect( Collectors.toMap( id -> id, id -> storage.data.get( id ) ) );
+//
+//        collection.updateMany( all( "_id", data.keySet() ), new BsonDocument( data.values() ) )
+    }
+
+    @Override
+    public void updated( Collection<T> objects, boolean added ) {
+        if( added ) {
+            Threads.synchronously( lock, () -> {
+                List<Metadata<T>> metadataValues = objects.stream()
+                    .map( object -> storage.data.get( storage.identifier.get( object ) ) )
+                    .collect( Collectors.toList() );
+                collection.insertMany( metadataValues );
+            } );
+        } else {
+            updated( objects );
+        }
+    }
+
+    @Override
+    public void deleted( T object ) {
+        Threads.synchronously( lock, () -> collection.deleteOne( eq( "_id", storage.identifier.get( object ) ) ) );
+    }
+
+    @Override
+    public void deleted( Collection<T> objects ) {
+        objects.forEach( this::deleted );
+        // TODO: rewrite with collection.deleteMany( eq( values ) );
+    }
+
+    @Override
+    public void fsync() {
+        Threads.synchronously( lock, () -> fsync( scheduled.lastExecuted() ) );
+    }
+
+    /**
+     * @param last executed sync
+     * TODO: avoid explicit usage of fsync
+     */
+    private void fsync( long last ) {
+        Threads.synchronously( lock, () -> {
+            log.trace( "fsyncing, last: {}, storage length: {}", last, storage.data.size() );
+            for( var value : storage.data.values() ) {
+                if( value.modified >= last ) persist( value );
```

Comment:
This implementation persists only insertions. The previous implementation has been persisting all updates. Fsync is the only point of persistence (except deletions). Have a look at MongoStorage fsync.

Reply:
That's true. I made appropriate changes.

The only difference with `MongoStorage` that there `Storage`.`fsync` is used, but in `MongoPersistence` - `DataListener`.`fsync` is used. There is a comment, that we shouldn't rely on `MongoStorage`.`fsync` explicitly and this method should be removed. So do we still want to keep it along with `DataListener`.`fsync`?